tools/build-flow-types.js

```diff
@@ -0,0 +1,155 @@
+// @flow
+/* eslint-disable no-console, import/no-commonjs */
+/**
+ * Build flow type files for exports.
+ *
+ * Generates *.js.flow files in dist for each file that is an export for
+ * a given package. It means we have to ship source code but we're doing that
+ * anyway right now, so no worries.
+ *
+ * Besides index.js files, we also export some other specific files for
+ * general consumption. In order to make this scalable, but reasonably fast
+ * we assume that each index.js file has a header in it that identifies
+ * additional export files. That way we only need to look for index.js files
+ * and read those to work out the rest. This makes it at least a little
+ * scalable but also self-documenting.
+ */
+const {promisify} = require("util");
+const fs = require("fs");
+const path = require("path");
+const ancesdir = require("ancesdir");
+
+/**
+ * Let's use async/await to make life nicer.
+ */
+const readFileAsync = promisify(fs.readFile);
+const writeFileAsync = promisify(fs.writeFile);
+
+/**
+ * Some regex's we use to support our additional-files syntax.
+ */
+/**
+ * The first extracts the @additional-exports block, which looks like this:
+ *
+ * @additional-exports [
+ *    "./filea.js",
+ *    "./fileb.js",
+ * ]
+ *
+ * Paths in that block are assumed to be relative to the directory of the file
+ * declaring them.
+ *
+ * The second regex grabs the filenames themselves from this block.
+ */
+const regexAdditionalFiles = /^(?:\s*.*@additional-exports\s?\[)((?:\s*.*"(?:[./\-\w]+)",\s*)*)(?:\s*.*\])$/im;
+const regexExtractFile = /^(?:\s*.*"([./\-\w]+)",\s*)$/gim;
```

Comment:
I'm going to assume that these are right.  My regex-fu is so bad.  😭

Reply:
I used regex101.com to create and test these. You could do the same to check them in a review.

Reply:
That's a good point.  If there are particular case I was worried about I could double check that it works with those.